Simplified/Reader2/NYPLReaderBookmarksBusinessLogic.swift

```diff
@@ -337,7 +344,105 @@ class NYPLReaderBookmarksBusinessLogic: NSObject, NYPLReadiumViewSyncManagerDele
     completion()
   }
 
-  // TODO: Sync reading position
+  // MARK: - Read Position
+  
+  func storeReadPosition(locator: Locator) {
+    // Avoid overwriting location when reader first open
+    guard (locator.locations.totalProgression ?? 0) != 0,
+      let bookLocation = NYPLBookLocation(locator: locator, publication: publication, renderer: NYPLBookLocation.r2Renderer) else {
+      return
+    }
+    
+    bookRegistryProvider.setLocation(bookLocation, forIdentifier: book.identifier)
+    postReadPosition(locationString: bookLocation.locationString)
+  }
+  
+  /// Restore locally stored location and  sync location from server
+  /// Both local and server fetch completions are only called if fetch succeed
+  func restoreReadPosition(currentLocator: Locator?, localFetchCompletion: (Locator?) -> (), serverFetchCompletion: @escaping (Locator?) -> ()) {
+    var currentLocation:NYPLBookLocation?
+    
+    if let currentLocator = currentLocator {
+      currentLocation = NYPLBookLocation(locator: currentLocator, publication: publication, renderer: NYPLBookLocation.r2Renderer)
+    }
+    
+    if let lastSavedLocation = bookRegistryProvider.location(forIdentifier: book.identifier),
+      let locator = lastSavedLocation.convertToLocator() {
+      localFetchCompletion(locator)
+      currentLocation = lastSavedLocation
+    }
+    
+    syncReadPosition(bookID: book.identifier, currentLocation: currentLocation, url: book.annotationsURL) { (serverLocationString) in
+      guard let locationString = serverLocationString,
+        let locator = NYPLBookLocation(locationString: locationString, renderer: NYPLBookLocation.r2Renderer).convertToLocator() else {
+        Log.debug(#file, "Unable to create locator from server stored location string - \(serverLocationString ?? "")")
+        return
+      }
+      
+      serverFetchCompletion(locator)
+    }
+  }
+  
+  /// Post the read position to server if we have synced the read position from server
+  /// There is a 120 seconds interval to avoid high frequency of requests
+  /// - Parameters:
+  /// - locationString: A json string that contains required information to create a Locator object
+  private func postReadPosition(locationString: String) {
+    serialQueue.async {
+      self.queuedReadPosition = locationString
```

Comment:
sorry I missed this earlier. There's a retain cycle between self > serialQueue > async closure > self

Also below in `postQueuedReadPositionInSerialQueue`

Reply:
Fixed, thank you for catching it